### ISSUE TITLE
DivisionByZeroError: Division by zero in [...]/nextcloud/apps/maps/lib/Service/PhotofilesService.php:413

### ISSUE THREAD

**Issue by giterest** [nextcloud/maps#655]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps to reproduce
1. CLI > sudo -u www-data php -f [...]/nextcloud/cron.php
2.
3.

### Expected behaviour
Tell us what should happen - cron.php should run /wo error.

### Actual behaviour
Tell us what happens instead - cron.php not executed because of the error in the title.

### Server configuration

**Operating system:**
Linux 5.10.0-9-amd64 (Debian Bullseye)

**Web server:**
nginx/1.21.3

**Database:**
pgsql/13.4

**PHP version:**
php8.0-fpm (8.0.12)

**Nextcloud version:** (see Nextcloud admin page)
Nextcloud 22.2.0

**Updated from an older Nextcloud/ownCloud or fresh install:**
latest minor patch

**Where did you install Nextcloud from:**
web updater

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.8.0
  - activity: 2.15.0
  - bruteforcesettings: 2.2.0
  - calendar: 2.3.4
  - circles: 22.1.1
  - cloud_federation_api: 1.5.0
  - comments: 1.12.0
  - contacts: 4.0.3
  - contactsinteraction: 1.3.0
  - dashboard: 7.2.0
  - dav: 1.19.0
  - federatedfilesharing: 1.12.0
  - federation: 1.12.0
  - files: 1.17.0
  - files_pdfviewer: 2.3.0
  - files_rightclick: 1.1.0
  - files_sharing: 1.14.0
  - files_trashbin: 1.12.0
  - files_versions: 1.15.0
  - files_videoplayer: 1.11.0
  - firstrunwizard: 2.11.0
  - logreader: 2.7.0
  - lookup_server_connector: 1.10.0
  - maps: 0.1.9
  - nextcloud_announcements: 1.11.0
  - notes: 4.1.1
  - notifications: 2.10.1
  - oauth2: 1.10.0
  - password_policy: 1.12.0
  - photos: 1.4.0
  - previewgenerator: 3.1.1
  - privacy: 1.6.0
  - provisioning_api: 1.12.0
  - recommendations: 1.1.0
  - serverinfo: 1.12.0
  - settings: 1.4.0
  - sharebymail: 1.12.0
  - support: 1.5.0
  - survey_client: 1.10.0
  - systemtags: 1.12.0
  - tasks: 0.14.2
  - text: 3.3.0
  - theming: 1.13.0
  - twofactor_backupcodes: 1.11.0
  - twofactor_totp: 6.1.0
  - updatenotification: 1.12.0
  - user_status: 1.2.0
  - viewer: 1.6.0
  - weather_status: 1.2.0
  - workflowengine: 2.4.0
Disabled:
  - admin_audit
  - encryption
  - files_external
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***",
            "***REMOVED SENSITIVE VALUE***"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "pgsql",
        "version": "22.2.0.2",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "nextcloud_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "enable_previews": true,
        "maintenance": false,
        "theme": "",
        "log_type": "file",
        "logfile": "nextcloud.log",
        "logfilemode": 416,
        "loglevel": 2,
        "logtimezone": "Europe\/Vienna",
        "log_rotate_size": 10485760,
        "updater.release.channel": "stable",
        "default_phone_region": "AT",
        "app_install_overwrite": [
            "calendar",
            "previewgenerator"
        ],
        "twofactor_enforced": "true",
        "twofactor_enforced_groups": [],
        "twofactor_enforced_excluded_groups": [],
        "mail_smtpmode": "sendmail",
        "mail_sendmailmode": "smtp",
        "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***"
    }
}
```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...
SMB

**Are you using encryption:** yes/no
no

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
no

### Client configuration
**Browser:**
CLI (in that case)

**Operating system:**
WIN10

#### Web server error log
<details>
<summary>Web server error log</summary>

```
sudo -u www-data php -f [...]/nextcloud/cron.php
DivisionByZeroError: Division by zero in [...]/nextcloud/apps/maps/lib/Service/PhotofilesService.php:413
Stack trace:
#0 [...]/nextcloud/apps/maps/lib/Service/PhotofilesService.php(264): OCA\Maps\Service\PhotofilesService->getExif()
nextcloud/server#1 [...]/nextcloud/apps/maps/lib/BackgroundJob/AddPhotoJob.php(56): OCA\Maps\Service\PhotofilesService->addPhotoNow()
nextcloud/server#2 [...]/nextcloud/lib/public/BackgroundJob/Job.php(79): OCA\Maps\BackgroundJob\AddPhotoJob->run()
nextcloud/server#3 [...]/nextcloud/lib/public/BackgroundJob/QueuedJob.php(47): OCP\BackgroundJob\Job->execute()
nextcloud/server#4 [...]/nextcloud/cron.php(127): OCP\BackgroundJob\QueuedJob->execute()
nextcloud/server#5 {main}
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
n/a or irrelevant
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
n/a or irrelevant
```
</details>

**Comment by nidomiro** [comment]
Same as #645?

**Comment by nidomiro** [comment]
My cron is also not working with this plugin enabled and my log gets spammed with:
```
{"reqId":"KgPueafXprunGNtMGhHW","level":3,"time":"2021-12-27T10:25:02+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Division by zero","userAgent":"--","version":"22.2.3.0","exception":{"Exception":"DivisionByZeroError","Message":"Division by zero","Code":0,"Trace":[{"file":"/var/www/html/custom_apps/maps/lib/Service/PhotofilesService.php","line":264,"function":"getExif","class":"OCA\\Maps\\Service\\PhotofilesService","type":"->","args":[{"__class__":"OC\\Files\\Node\\File"}]},{"file":"/var/www/html/custom_apps/maps/lib/BackgroundJob/AddPhotoJob.php","line":56,"function":"addPhotoNow","class":"OCA\\Maps\\Service\\PhotofilesService","type":"->","args":[{"__class__":"OC\\Files\\Node\\File"},"******"]},{"file":"/var/www/html/lib/public/BackgroundJob/Job.php","line":79,"function":"run","class":"OCA\\Maps\\BackgroundJob\\AddPhotoJob","type":"->","args":[{"photoId":331570,"userId":"******"}]},{"file":"/var/www/html/lib/public/BackgroundJob/QueuedJob.php","line":47,"function":"execute","class":"OCP\\BackgroundJob\\Job","type":"->","args":[{"__class__":"OC\\BackgroundJob\\JobList"},{"__class__":"OC\\Log"}]},{"file":"/var/www/html/cron.php","line":127,"function":"execute","class":"OCP\\BackgroundJob\\QueuedJob","type":"->","args":[{"__class__":"OC\\BackgroundJob\\JobList"},{"__class__":"OC\\Log"}]}],"File":"/var/www/html/custom_apps/maps/lib/Service/PhotofilesService.php","Line":413,"CustomMessage":"--"},"id":"61c9951d93bdc"}
```

**Comment by kovge** [comment]
Hi

This pull request might resolve these issues: https://github.com/nextcloud/maps/pull/586
Do you have a chance to test it?

Thank you

**Comment by klaHoadl** [comment]
Hello, 

I had the same issue with:
`/var/www/html/custom_apps/maps/lib/Service/PhotofilesService.php, line 264.`

Quick fix was to surround the line with a try/catch blog:

```
$exif=null;
try{
        $exif = $this->getExif($photo);

}catch(\Throwable $f){
        $this->logger->error($f->getMessage());
}
```

**Comment by kovge** [comment]
> 

Hi @klaHoadl ,

Here: https://github.com/nextcloud/maps/pull/586/files#diff-f30091dfcfc080a0ac3527ad223fd414874701d90f7e084b70c3dcd832410976R365 Happened a simpilar thing.

Your solution has one issue, it catches some errors as well, witch are produced by incompatibility of php7.4 and upper.
For example previously the code did not thrown an error if you did something like that:
`$x = '5'; $y = $x / 5;`
But as php, expecially with strict standards enabled, throws errors in the case i wrote here.
These are catched by you code, but did not solved the original problem. The same thing with division by 0 errors.
In one hand your code will run through the images, but will not process all valid metada, and you will have just a handful of succesfully processed images. So summa-summarum:  the try-catch just stops the error to be propagated to upper levels, and prevents the process to be stopped, but the error will still be there.

In the corresponding pull request, which i mentioned previously: https://github.com/nextcloud/maps/pull/586/files#diff-b61e1377ced1f6a68d605a0036826d010309ea9408d1d261dce56a9d1149a63a
There are a bunch of type checks and typecasts to prevent the problems ,like the one, caused the issue in your case. It checks 0 values as well, to prevent possible division by 0 errors. Unfortunately regarding the (different format and quality of) metadata of the different image files, there could be valid data there, just not processable with the actual production code.

**Comment by klaHoadl** [comment]
Hello @kovge ,

thanks for your explanation! 

I'm only little familiar with PHP and github therefore I was looking for an opportunity to process all pictures to have them on the map. I also do not know how to set up the pull requested solution on my setup, so I ended up in the quick fix above.

**Comment by tacruc** [comment]
Closed by #586
